src/core/net/dns_dso.cpp

```diff
@@ -0,0 +1,1537 @@
+
+    case kStateEstablishingSession:
+        // Unidirectional message are allowed after session is
+        // established. While session is being established, on client,
+        // we allow rx on response message. On server we can rx
+        // request or response.
+
+        VerifyOrExit(header.GetMessageId() != 0);
+
+        if (IsClient())
+        {
+            VerifyOrExit(header.GetType() == Dns::Header::kTypeResponse);
+        }
+        break;
+
+    case kStateSessionEstablished:
+        // All message types are allowed.
+        break;
+
+    case kStateDisconnected:
+    case kStateConnecting:
+        ExitNow();
+    }
+
+    // All count fields MUST be set to zero in the header.
+    VerifyOrExit((header.GetQuestionCount() == 0) && (header.GetAnswerCount() == 0) &&
+                 (header.GetAuthorityRecordCount() == 0) && (header.GetAdditionalRecordCount() == 0));
+
+    aMessage.SetOffset(sizeof(header));
+
+    switch (ReadPrimaryTlv(aMessage, primaryTlvType))
+    {
+    case kErrorNone:
+        VerifyOrExit(primaryTlvType != Tlv::kReservedType);
+        break;
+
+    case kErrorNotFound:
+        // The `primaryTlvType` is set to `Tlv::kReservedType`
+        // (value zero) to indicate that there is no primary TLV.
+        break;
+
+    default:
+        ExitNow();
+    }
+
+    switch (header.GetType())
+    {
+    case Dns::Header::kTypeQuery:
+        error = ProcessRequestOrUnidirectionalMessage(header, aMessage, primaryTlvType);
+        break;
+
+    case Dns::Header::kTypeResponse:
+        error = ProcessResponseMessage(header, aMessage, primaryTlvType);
+        break;
+    }
+
+exit:
+    aMessage.Free();
+
+    if (error == kErrorNone)
+    {
+        ResetTimeouts(/* aIsKeepAliveMessage */ (primaryTlvType == KeepAliveTlv::kType));
+    }
+    else
+    {
+        Disconnect(kForciblyAbort, kReasonPeerMisbehavior);
+    }
+
+    // We signal any state change at the very end when the received
+    // message is fully processed (all state and local variables are
+    // updated) to ensure that we do not have any reentrancy issues
+    // (e.g., if a `Connection` method happens to be called from the
+    // callback).
+
+    SignalAnyStateChange();
+}
+
+Error Dso::Connection::ReadPrimaryTlv(const Message &aMessage, Tlv::Type &aPrimaryTlvType) const
+{
+    // Read and validate the primary TLV (first TLV  after the header).
+    // The `aMessage.GetOffset()` must point to the first TLV. If no
+    // TLV then `kErrorNotFound` is returned. If TLV in message is not
+    // well-formed `kErrorParse` is returned. The read TLV type is
+    // returned in `aPrimaryTlvType` (set to `Tlv::kReservedType`
+    // (value zero) when `kErrorNotFound`).
+
+    Error error = kErrorNotFound;
+    Tlv   tlv;
+
+    aPrimaryTlvType = Tlv::kReservedType;
+
+    SuccessOrExit(aMessage.Read(aMessage.GetOffset(), tlv));
+    VerifyOrExit(aMessage.GetOffset() + tlv.GetSize() <= aMessage.GetLength(), error = kErrorParse);
```

Comment:
Do we need to worry about potential overflow of `offset + size`?

Reply:
I think it is covered and handled properly.
-  The DSO TLVs use `uint16_t` as length (`Tlv::GetLenhth()` returns `uint16_t`).
- But `Tlv::GetSize()` actually returns `uint32_t` (to ensure size is correct even when length is `UINT16_MAX`).
- So the check here is done in `uint32_t` which should be fine.